### DISCUSSION TITLE
Error while using exported  .pb model

### DISCUSSION THREAD

**Discussion by Karshiev** [ultralytics/yolov5#10626] [Q&A] [answered]
Recently I trained yolov5x model to detect fruits and vegetables dataset. 
After converting the trained yolov5x.pt file into .pb file, I loaded the .pb model and tried to get classes and detected objects with the following code:

```
**import skimage.io
model_path = r'path_to_.pb_model'
loaded_model = tf.saved_model.load(model_path)
IMAGE_PATH = r'./test_image.jpg'
image = skimage.io.imread(IMAGE_PATH, as_gray=False)
image_expanded = np.expand_dims(image, axis=0)
output = loaded_model(image_expanded)
boxes = np.squeeze(output['detection_boxes'])**

```
I had following issue: 

Traceback (most recent call last):
  File "test_pb.py", line 38, in <module>
    boxes = np.squeeze(output['detection_boxes'])
**TypeError: list indices must be integers or slices, not str**

It seems the output is "list", but it must be dictionary.
I am new to yolov5, Can you please help me?

**Comment by alijafari79** [comment, marked as answer]
> Recently I trained yolov5x model to detect fruits and vegetables dataset. After converting the trained yolov5x.pt file into .pb file, I loaded the .pb model and tried to get classes and detected objects with the following code:
> 
> ```
> **import skimage.io
> model_path = r'path_to_.pb_model'
> loaded_model = tf.saved_model.load(model_path)
> IMAGE_PATH = r'./test_image.jpg'
> image = skimage.io.imread(IMAGE_PATH, as_gray=False)
> image_expanded = np.expand_dims(image, axis=0)
> output = loaded_model(image_expanded)
> boxes = np.squeeze(output['detection_boxes'])**
> ```
> 
> I had following issue:
> 
> Traceback (most recent call last): File "test_pb.py", line 38, in boxes = np.squeeze(output['detection_boxes']) **TypeError: list indices must be integers or slices, not str**
> 
> It seems the output is "list", but it must be dictionary. I am new to yolov5, Can you please help me?

### Hi @Karshiev, 

I got what you say and let's do it this way to get the keys you need :)

First let's see how the model organizes the keys ....

#### At this step load the model :

```
model_path = r'path_to_.pb_model'
loaded_model = tf.saved_model.load(model_path)
````

Then, Let's see the keys using signatures :

```
print(list(loaded_model .signatures.keys()))
```

Imported signatures always return dictionaries.

Then for the output of the last command you may see the keys of the model signature e.g :

``` ['predictions'] ```

Now for the rest of the code, I think the solution is to give the "loaded_model" the input data and then see the "signatures" of the "output" that will be a dictionary and will shed light on what you may look for in the future :

The rest of the code like you wrote :
```
IMAGE_PATH = r'./test_image.jpg'
image = skimage.io.imread(IMAGE_PATH, as_gray=False)
image_expanded = np.expand_dims(image, axis=0)
```

but a change here to see the keys of the output :

```
output = loaded_model(image_expanded)
print(list(output.signatures.keys()))
```

and you may see the result that you look for :)

Dear @Karshiev , if you found my answer useful **Mark it as answer** !

**Reply by alijafari79** [reply]
Did the solution above work for you ??

**Comment by alijafari79** [comment]
> Recently I trained yolov5x model to detect fruits and vegetables dataset. After converting the trained yolov5x.pt file into .pb file, I loaded the .pb model and tried to get classes and detected objects with the following code:
> 
> ```
> **import skimage.io
> model_path = r'path_to_.pb_model'
> loaded_model = tf.saved_model.load(model_path)
> IMAGE_PATH = r'./test_image.jpg'
> image = skimage.io.imread(IMAGE_PATH, as_gray=False)
> image_expanded = np.expand_dims(image, axis=0)
> output = loaded_model(image_expanded)
> boxes = np.squeeze(output['detection_boxes'])**
> ```
> 
> I had following issue:
> 
> Traceback (most recent call last): File "test_pb.py", line 38, in boxes = np.squeeze(output['detection_boxes']) **TypeError: list indices must be integers or slices, not str**
> 
> It seems the output is "list", but it must be dictionary. I am new to yolov5, Can you please help me?

Did the solution above work for you ??

**Reply by Karshiev** [reply]
Sorry for the late responce. 
The solution you provided is only for this code. But, we were working with tensorflow servings. It seems the solution you have provided is not the way I was looking for.  We have already found the solution to make corrections in the code in the model deployment, not in the base code.

**Reply by Karshiev** [reply]
Anyway, thank you for your responce.
